src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs

```diff
@@ -2811,15 +3013,108 @@ public static class GenExtensions<X>
+                // (19,35): error CS81250:  The parameter modifier 'out' cannot be used with 'ref' 
+                //     public static void Foo<T>(ref out X x) {}
+                Diagnostic(ErrorCode.ERR_BadParameterModifiers, "out").WithArguments("out", "ref").WithLocation(19, 35),
+                // (21,39): error CS81250:  The parameter modifier 'out' cannot be used with 'ref' 
+                //     public static void Foo<T,U,V>(ref out U u) {}
+                Diagnostic(ErrorCode.ERR_BadParameterModifiers, "out").WithArguments("out", "ref").WithLocation(21, 39),
+                // (22,39): error CS81250:  The parameter modifier 'out' cannot be used with 'ref' 
+                //     public static void Foo<T,U,V>(ref out X x) {}
+                Diagnostic(ErrorCode.ERR_BadParameterModifiers, "out").WithArguments("out", "ref").WithLocation(22, 39));
+        }
+
+        [Fact]
+        public void DuplicateParameterModifiersWillErrorOut()
+        {
+            var test = @"
+public static class TestType
+{
+    public static void Test(ref ref int i) {}
+    public static void Test(out out int i) {}
+    public static void Test(in in int i) {}
+    public static void Test(this this int i) {}
+    public static void Test(params params int[] i) {}
+}
+";
+
+            ParseAndValidate(test,
+                new CSharpParseOptions(LanguageVersion.CSharp7_1),
+                // (4,33): error CS1107: A parameter can only have one 'ref' modifier
+                //     public static void Test(ref ref int i) {}
+                Diagnostic(ErrorCode.ERR_DupParamMod, "ref").WithArguments("ref").WithLocation(4, 33),
+                // (5,33): error CS1107: A parameter can only have one 'out' modifier
+                //     public static void Test(out out int i) {}
+                Diagnostic(ErrorCode.ERR_DupParamMod, "out").WithArguments("out").WithLocation(5, 33),
+                // (6,32): error CS1107: A parameter can only have one 'in' modifier
+                //     public static void Test(in in int i) {}
+                Diagnostic(ErrorCode.ERR_DupParamMod, "in").WithArguments("in").WithLocation(6, 32),
+                // (7,34): error CS1107: A parameter can only have one 'this' modifier
+                //     public static void Test(this this int i) {}
+                Diagnostic(ErrorCode.ERR_DupParamMod, "this").WithArguments("this").WithLocation(7, 34),
+                // (8,36): error CS1107: A parameter can only have one 'params' modifier
+                //     public static void Test(params params int[] i) {}
+                Diagnostic(ErrorCode.ERR_DupParamMod, "params").WithArguments("params").WithLocation(8, 36));
+        }
+
+        [Fact]
+        public void InParametersAreParsedCorrectly()
+        {
+            var test = @"
+public static class Test
+{
+    public delegate void Delegate(in int a);
+
+    public static void Method(in int b)
+    {
+        void LocalFunc(in int c) { }
+    }
```

Comment:
lambdas are particularly interesting, since they parse parameters differently.

---
In reply to: [101646598](https://github.com/dotnet/roslyn/pull/17187#discussion_r101646598) [](ancestors = 101646598)